res/controllers/midi-components-0.1.js

```diff
@@ -0,0 +1,1135 @@
+        constructor(options) {
+            if (options.number === undefined) {
+                throw "No hotcue number specified for HotcueButton constructor";
+            }
+            if (options.colorMapper !== undefined || options.sendRGB !== undefined) {
+                options.colorKey = "hotcue_" + options.number + "_color";
+            }
+            options.outKey = "hotcue_" + options.number + "_enabled";
+            super(options);
+        }
+        unshift() {
+            this.inKey = "hotcue_" + this.number + "_activate";
+        }
+        shift() {
+            this.inKey = "hotcue_" + this.number + "_clear";
+        }
+        output(value) {
+            const outval = this.outValueScale(value);
+            // NOTE: outputColor only handles hotcueColors
+            // and there is no hotcueColor for turning the LED
+            // off. So the `send()` function is responsible for turning the
+            // actual LED off.
+            if (this.colorKey !== undefined && outval !== this.off) {
+                this.outputColor(engine.getValue(this.group, this.colorKey));
+            } else {
+                this.send(outval);
+            }
+        }
+        outputColor(colorCode) {
+            // Sends the color from the colorCode to the controller. This
+            // method will not be called if no colorKey has been specified.
+            if (colorCode === undefined || colorCode < 0 || colorCode > 0xFFFFFF) {
+                console.warn("Ignoring invalid color code '" + colorCode + "' in outputColor()");
+                return;
+            }
+
+            if (this.colorMapper !== undefined) {
+                // This HotcueButton holds a reference to a ColorMapper. This means
+                // that the controller only supports a fixed set of colors, so we
+                // get the MIDI value for the nearest supported color and send it.
+                const nearestColorValue = this.colorMapper.getValueForNearestColor(colorCode);
+                this.send(nearestColorValue);
+            } else {
+                // Since outputColor has been called but no ColorMapper is
+                // available, we can assume that controller supports arbitrary
+                // RGB color output.
+                this.sendRGB(colorCodeToObject(colorCode));
+            }
+        }
+        sendRGB(_colorObject) {
+            // This method needs to be overridden in controller mappings,
+            // because the procedure is controller-dependent.
+            throw "sendRGB(colorObject) not implemented -- unable to send RGB colors";
+        }
+        connect() {
+            super.connect();
+            if (this.group !== undefined && this.colorKey !== undefined) {
+                // TODO (Swiftb0y): replace with arrow function once https://bugreports.qt.io/browse/QTBUG-95677 got fixed
+                this.connections[1] = engine.makeConnection(this.group, this.colorKey, function(color) {
```

Comment:
does this override connection number two, or is it meant to append? This seems like a dangerous magic number.

Reply:
`this.connections` is more like an array with slots for storing connections. IMO its a questionable design choice but I'm not sure how else to implement. Maybe a Symbols could help.

Reply:
It is meant to override. If you want to change this implementation detail to use `Array.push`, I think you'd need to clear the Array first.

Reply:
Don't we need to disconnect the original connection before we overwrite it? Otherwise we've got a dangling connection that won't get removed if the script is reloaded.

Reply:
It is the caller's responsibility to call `disconnect` before `connect`.

Reply:
Mhmm. we might want to introduce an abstraction for the `connections` then.

Reply:
I'm not sure what you mean. `disconnect` and `connect` are the abstractions for the callee. The `connections` array is not intended to be directly manipulated from the outside (and JavaScript has no way to enforce that...).

Reply:
The problem is that we need some way to talk about a connectionHandler (returned from `engine.makeConnection`) while being able to iterate over all owned by a particular object. I think @ywwg is concerned that these are currently just addressed by their index. We might instead want to consider making `connections` a `Map` or an `Object` so we can address these handlers using something more readable.
Would just using a plain `Object` suffice?

Reply:
I am ok with making this a TODO as well -- we should come up with some way of handling connections that doesn't rely so much on callers doing the right thing.